### PR TITLE
Metricbeat elasticsearch/node_stats metricset: adding more metrics

### LINKED ISSUE #7848: Metricbeat - Elasticsearch module - missing some node metrics
For confirmed bugs, please report:
- Version: 6.4 
- Operating System: OSX

- Steps to Reproduce: 
Deploy elasticsearch cluster with metricbeat over K8S. 
Configure metricbeat to report metrics to elasticsearch and look at metricbeat index in kibana 
The following metricsets are missing comapred to [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html):

* Operating system stats, load average, mem, swap
* Process statistics, memory consumption, cpu usage, open file descriptors (except for mlockall which is reported)
* Statistics about each thread pool, including current size, queue and rejected tasks
* Transport statistics about sent and received bytes in cluster communication
* breaker - Statistics about the field data circuit breaker
* Discovery - Statistics about the discovery
* Ingest - Statistics about ingest preprocessing
* adaptive_selection - Statistics about adaptive replica selection

### PR DESCRIPTION
**Pull request by ycombinator**
Resolves #7848 

This PR adds the following metrics to the `elasticsearch/node_stats` metricset:
* [x] Statistics about each thread pool, including current size, queue and rejected tasks
* [x] Transport statistics about sent and received bytes in cluster communication
* [x] breaker - Statistics about the field data circuit breaker
* [x] Discovery - Statistics about the discovery
* [x] Ingest - Statistics about ingest preprocessing
* [x] adaptive_selection - Statistics about adaptive replica selection

Also TODO:
* [ ] Fix CI errors

### CODE REVIEW

File: metricbeat/module/elasticsearch/node_stats/_meta/fields.yml [hunk 1]
```diff
@@ -156,3 +156,154 @@
...
+            - name: current
+              type: long
+              description: >
+            - name: failed
+              type: long
+              description: >
+            - name: ms
+              type: long
+              description: >
+        - name: pipelines.*
```
Comment (ycombinator) [inline metricbeat/module/elasticsearch/node_stats/_meta/fields.yml]: @ruflin I'm not sure the section for this field is generating the correct ES template. Could you take a look at the generated template and and the sample data for this field in `data.json` and tell me how I can generate the correct template?

File: metricbeat/module/elasticsearch/node_stats/data.go [hunk 2]
```diff
@@ -123,6 +199,14 @@ func eventsMapping(r mb.ReporterV2, content []byte) error {
 			r.Error(err)
 		}
 
+		// ingest.pipelines.*
+		value, ok := node["ingest"].(map[string]interface{})["pipelines"]
```
Comment (ycombinator) [inline metricbeat/module/elasticsearch/node_stats/data.go]: @ruflin Is this multi-level indexing into the map + type assertion, all in one step, safe to do or should I break this up into 3 separate steps (index into "ingest", then the type assertion, then index into "pipelines") with error checking after each step?

File: metricbeat/module/elasticsearch/node_stats/data.go [hunk 3]
```diff
@@ -98,13 +165,48 @@ var (
...
+	}
 )
 
 type nodesStruct struct {
 	ClusterName string                            `json:"cluster_name"`
 	Nodes       map[string]map[string]interface{} `json:"nodes"`
 }
 
+func pipelineMetricsMapping(node map[string]interface{}, metricSetFields common.MapStr, r mb.ReporterV2) {
+	value, ok := node["ingest"].(map[string]interface{})["pipelines"]
```
Comment (ycombinator) [inline metricbeat/module/elasticsearch/node_stats/data.go]: @ruflin Is this multi-level indexing into the map + type assertion, all in one step, safe to do or should I break this up into 3 separate steps (index into "ingest", then the type assertion, then index into "pipelines") with error checking after each step?

File: metricbeat/module/elasticsearch/node_stats/data.go [hunk 4]
```diff
@@ -98,13 +160,84 @@ var (
...
+	}
 )
 
 type nodesStruct struct {
 	ClusterName string                            `json:"cluster_name"`
 	Nodes       map[string]map[string]interface{} `json:"nodes"`
 }
 
+func pipelineMetricsMapping(node map[string]interface{}, metricSetFields common.MapStr, r mb.ReporterV2) {
+	pipelines, ok := node["ingest"].(map[string]interface{})["pipelines"].(map[string]interface{})
```
Comment (ycombinator) [inline metricbeat/module/elasticsearch/node_stats/data.go]: @ruflin Is this multi-level indexing into the map + type assertion, all in one step, safe to do or should I break this up into separate steps with error checking after each step?

**Comment by ruflin** [comment]
@ycombinator Let's first continue the discussion in #7848 to decide which metrics we should actually add. Just adding all metrics there are I don't think will help us in long run.

**Comment by ycombinator** [comment]
@ruflin I think we should close this PR unmerged and continue the discussion on the issue (#7848). We can always resurrect the PR from the branch if we want later. WDYT?

**Pull request closed without merging**